src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java

```diff
@@ -84,26 +88,114 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
+              .thenApply(r -> r.next(items -> segregateItemsList(items, requestRelatedRecords)));
+
+    CompletableFuture<Result<InstanceRequestRelatedRecords>> sortedAvailableItemsFuture = itemsListsFuture.thenApply(r -> r.next(RequestByInstanceIdResource::rankItemsByMatchingServicePoint));
+    CompletableFuture<Result<Map<Item, DateTime>>> loanItemsFuture = itemsListsFuture.thenCompose(r -> r.after(relatedRecords -> getLoanItems(relatedRecords, clients)));
+    CompletableFuture<Result<Map<Item, Integer>>> requestQueuesFuture = itemsListsFuture.thenCompose(r -> r.after(relatedRecords -> getRequestQueues(relatedRecords, clients)));
+
+    CompletableFuture.allOf(sortedAvailableItemsFuture, loanItemsFuture, requestQueuesFuture)
+                     .thenCompose( x -> CompletableFuture.completedFuture(
+                          combineWithUnavailableItems(loanItemsFuture.join(), requestQueuesFuture.join(), requestRelatedRecords)))
+                     .thenApply( r -> r.next(RequestByInstanceIdResource::instanceToItemRequests))
+                     .thenCompose( r -> r.after( requests -> placeRequests(requests, clients)))
+                     .thenApply(r -> r.map(RequestAndRelatedRecords::getRequest))
+                     .thenApply(r -> r.map(new RequestRepresentation()::extendedRepresentation))
+                     .thenApply(CreatedJsonResponseResult::from)
+                     .thenAccept(result -> result.writeTo(routingContext.response()))
+                     .exceptionally( err -> {
+                         String reason = "Error processing title-level request";
+                         log.error(reason, err);
+                         ServerErrorResponse.internalError(routingContext.response(), reason);
+                         return null;
+                      });
+  }
+
+  private CompletableFuture<Result<Map<Item, DateTime>>> getLoanItems(
+    InstanceRequestRelatedRecords instanceRequestPackage, Clients clients) {
+
+    final List<Item> unsortedUnavailableItems = instanceRequestPackage.getUnsortedUnavailableItems();
+    if (unsortedUnavailableItems == null ||
+        unsortedUnavailableItems.isEmpty()) {
+      return CompletableFuture.completedFuture(succeeded(null));
+    }
+
+    LoanRepository loanRepository = new LoanRepository(clients);
+    Map<Item, CompletableFuture<Result<Loan>>> itemLoanFuturesMap = new HashMap<>();
+
+    //Find request queues and loan items for each item
+    for (Item item : unsortedUnavailableItems) {
+      itemLoanFuturesMap.put(item, loanRepository.findOpenLoanForItem(item));
+    }
+
+    final Collection<CompletableFuture<Result<Loan>>> loanFutures = itemLoanFuturesMap.values();
+
+    return CompletableFuture.allOf(loanFutures.toArray(new CompletableFuture[0]))
+      .thenApply(dd -> {
+        Map<Item, DateTime> itemDueDateMap = new HashMap<>();
+        List<Item> itemsWithoutLoansList = new LinkedList<>(); //to preserve the order it was found.
```

Comment:
I don't get why the order found needs to be preserved here. The order should not be predictable as we work through the `itemLoanFuturesMap` below.

Reply:
Probably leftover from previous implementation when order was required. Comment needs updating as well. Coming from .NET where the list's order is preserved by default, I feel "safer" with an ordered list. After making the change to ArrayList the tests still pass so I'll keep it with ArrayList.